File: VoucherManagementBE/src/main/java/com/example/vouchermanagementbe/module/voucher/domain/entity/Voucher.java
package com.example.vouchermanagementbe.module.voucher.domain.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Voucher {
    private Long id;
    private String code;
    private Integer discountPercent;
    private Integer quantity;
    private LocalDate expiredDate;
    private VoucherStatus status;
    private LocalDateTime createdAt;

    // Logic khởi tạo
    public void validateForCreation() {
        if (discountPercent == null || discountPercent < 1 || discountPercent > 100) {
            throw new RuntimeException("Discount percent must be between 1 and 100");
        }
        if (quantity == null || quantity < 0) {
            throw new RuntimeException("Quantity must be greater than or equal to 0");
        }
        if (expiredDate == null || expiredDate.isBefore(LocalDate.now()) || expiredDate.isEqual(LocalDate.now())) {
            throw new RuntimeException("Expired date must be in the future");
        }
    }

    // Nghiệp vụ sử dụng voucher
    public void use() {
        if (this.status == VoucherStatus.INACTIVE) {
            throw new RuntimeException("Voucher is INACTIVE");
        }
        if (this.expiredDate.isBefore(LocalDate.now())) {
            throw new RuntimeException("Voucher is expired");
        }
        if (this.quantity <= 0) {
            throw new RuntimeException("Voucher is out of stock");
        }
        this.quantity -= 1;
    }
}

File: VoucherManagementFE/src/core/domain/entities/Voucher.ts
/**
 * Voucher Entity - Represents a promotional voucher in the system
 * Matches API response: { code, createdAt, discountPercent, expiredDate, id, quantity, status }
 */
export class Voucher {
  id?: number;
  code: string;
  discountPercent: number;
  quantity: number;
  expiredDate: string; // ISO 8601 format: YYYY-MM-DD
  status: string; // e.g., "ACTIVE", "EXPIRED", "INACTIVE"
  createdAt?: string;

  constructor(
    code: string,
    discountPercent: number,
    quantity: number,
    expiredDate: string,
    status: string = "ACTIVE",
    id?: number,
    createdAt?: string
  ) {
    this.id = id;
    this.code = code;
    this.discountPercent = discountPercent;
    this.quantity = quantity;
    this.expiredDate = expiredDate;
    this.status = status;
    this.createdAt = createdAt;
  }

  /**
   * Validate voucher business rules
   * Returns array of validation errors (empty if valid)
   */
  validate(): string[] {
    const errors: string[] = [];

    if (!this.code?.trim()) {
      errors.push("Code is required");
    }

    if (this.discountPercent < 1 || this.discountPercent > 100) {
      errors.push("Discount percent must be between 1 and 100");
    }

    if (this.quantity < 0) {
      errors.push("Quantity must be greater than or equal to 0");
    }

    const expiredDate = new Date(this.expiredDate);
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    if (expiredDate <= today) {
      errors.push("Expired date must be greater than today");
    }

    return errors;
  }

  /**
   * Check if voucher is still valid (not expired)
   */
  isValid(): boolean {
    const expiredDate = new Date(this.expiredDate);
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    return expiredDate > today && this.status === "ACTIVE";
  }

  /**
   * Check if voucher has sufficient quantity
   */
  hasAvailableQuantity(): boolean {
    return this.quantity > 0;
  }
}
